main/src/ca/uwaterloo/flix/language/ast/UnkindedType.scala

```diff
@@ -0,0 +1,595 @@
+/*
+ * Copyright 2021 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.ast
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+/**
+  * A type without an associated kind.
+  */
+sealed trait UnkindedType {
+
+  /**
+    * Returns the type constructor of the given type.
+    */
+  def typeConstructor: Option[UnkindedType.Constructor] = this match {
+    case UnkindedType.Cst(cst, _) => Some(cst)
+    case UnkindedType.Apply(t1, _) => t1.typeConstructor
+    case UnkindedType.Lambda(_, _) => throw InternalCompilerException("Unexpected type constructor: Lambda")
```

Comment:
We usually move the exception cases to the end of the match (in other phases we do this a lot)